### PR TITLE
DBZ-688 Support for test execution on MySQL 8 server

### PR DESCRIPTION
**Pull request by jpechane**
There are 5 test failures under investigation

### CODE REVIEW

File: jenkins-jobs/mysql-matrix-test.yaml [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# Job definition to test MySQL connector against different MySQL versions
```
Comment (gunnarmorling) [inline jenkins-jobs/mysql-matrix-test.yaml]: I'm not sure whether/where we CI this already, but can we add GTID on/off as an axis here, too?

**Comment by jpechane** [comment]
@gunnarmorling All changes should be done.

File: jenkins-jobs/mysql-matrix-test.yaml (hunk 1 above)
Comment (jpechane) [reply jenkins-jobs/mysql-matrix-test.yaml]: @gunnarmorling I've enabled `assembly` profile that should execute tests against multiple topologies.

File: debezium-connector-mysql/src/test/docker/init-alt-replica/setup.sql [hunk 2]
```diff
@@ -6,12 +6,14 @@
 -- 2) 'snapper' - all privileges required by the snapshot reader AND binlog reader
 -- 3) 'mysqluser' - all privileges
 --
-GRANT REPLICATION SLAVE, REPLICATION CLIENT ON *.* TO 'replicator' IDENTIFIED BY 'replpass';
-GRANT SELECT, RELOAD, SHOW DATABASES, REPLICATION SLAVE, REPLICATION CLIENT  ON *.* TO 'snapper'@'%' IDENTIFIED BY 'snapperpass';
+CREATE USER 'replicator' IDENTIFIED BY 'replpass';
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/docker/init-alt-replica/setup.sql]: @jpechane Why are these changes needed? Should we add something to the docs, too, regarding permissions needed by the connector?

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MetadataIT.java [hunk 3]
```diff
@@ -135,7 +135,8 @@ public void shouldLoadMetadataViaJdbc() throws SQLException {
             assertFalse(product.columnWithName("createdByDate").scale().isPresent());
             assertThat(product.columnWithName("createdByDate").position()).isEqualTo(2);
             assertThat(product.columnWithName("createdByDate").isAutoIncremented()).isFalse();
-            assertThat(product.columnWithName("createdByDate").isGenerated()).isFalse();
+            // Different behaviour between MySQL 5.7 and MySQL 8
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MetadataIT.java]: Can you expand the comment a bit; what's is the difference exactly?

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlGeometryIT.java [hunk 4]
```diff
@@ -72,12 +72,12 @@ public void shouldConsumeAllEventsFromDatabaseUsingBinlogAndNoSnapshot() throws
         //Testing.Debug.enable();
         int numCreateDatabase = 1;
         int numCreateTables = 2;
-        int numDataRecords = 4 + 2;
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySqlGeometryIT.java]: Can you elaborate a bit on this change?

File: debezium-connector-mysql/src/test/resources/ddl/geometry_test.sql [hunk 5]
```diff
@@ -16,9 +16,10 @@ CREATE TABLE dbz_222_point (
   expected_srid INT,
   PRIMARY KEY (id)
 ) DEFAULT CHARSET=utf8;
-INSERT INTO dbz_222_point VALUES (default,GeomFromText('POINT(1 1)'), 1.0, 1.0, NULL);
-INSERT INTO dbz_222_point VALUES (default,GeomFromText('POINT(8.25554554 3.22124447)', 4326), 8.25554554, 3.22124447, 4326);
-INSERT INTO dbz_222_point VALUES (default,GeomFromText('POINT(0 0)', 1234), 0.0, 0.0, 1234);
+INSERT INTO dbz_222_point VALUES (default,ST_GeomFromText('POINT(1 1)'), 1.0, 1.0, NULL);
+INSERT INTO dbz_222_point VALUES (default,ST_GeomFromText('POINT(8.25554554 3.22124447)', 4326), 8.25554554, 3.22124447, 4326);
+-- Invalid SRID is not supported in MySQL 8
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/resources/ddl/geometry_test.sql]: Maybe this should be extracted into a separate file which isn't used with MySQL 8?

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MysqlDefaultValueIT.java [hunk 6]
```diff
@@ -517,7 +517,13 @@ public void dateAndTimeTest() throws InterruptedException {
         //current timestamp will be replaced with epoch timestamp
         ZonedDateTime t5 = ZonedDateTime.ofInstant(Instant.EPOCH, ZoneOffset.UTC);
         String isoString5 = ZonedTimestamp.toIsoString(t5, ZoneOffset.UTC, MySqlValueConverters::adjustTemporal);
-        assertThat(schemaJ.defaultValue()).isEqualTo(isoString5);
+        try {
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MysqlDefaultValueIT.java]: I'm wondering whether we can handle these subtle differences a bit more elegantly. How about injecting the DB version as an environment variable and then deciding based on that which assertion to run?

File: debezium-connector-mysql/src/test/docker/init-alt-replica/setup.sql (hunk 2 above)
Comment (jpechane) [reply debezium-connector-mysql/src/test/docker/init-alt-replica/setup.sql]: @gunnarmorling There is no difference in permitions - just the user can no longer be implictly created by `GRANT` but must be created exaplicitly.

**Comment by jpechane** [comment]
@gunnarmorling I've tried to extract the differences into a single central file

**Comment by gunnarmorling** [comment]
@jpechane Is there anything else to add, or is it good to be reviewed from your side?

**Comment by jpechane** [comment]
@gunnarmorling Good from my side

File: debezium-connector-mysql/src/test/docker/init-alt-replica/setup.sql (hunk 2 above)
Comment (gunnarmorling) [reply debezium-connector-mysql/src/test/docker/init-alt-replica/setup.sql]: Ah, yes. Thx!

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySQLConnection.java [hunk 7]
```diff
@@ -82,4 +88,85 @@ protected static void addDefaults(Configuration.Builder builder) {
...
+                        /**
+                         * MySQL 8 does not support unknown SRIDs so the case is removed
+                         */
+                        @Override
+                        public int geometryPointTableRecords() {
+                            return 3;
+                        }
+
+                        /**
+                         * MySQL 8 returns X and Y in a different order
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySQLConnection.java]: That sounds a bit scary. Is there code in the connector in place which keeps the existing order within emitted messages?

Comment (gunnarmorling) [reply debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/MySQLConnection.java]: Found [this post](https://mysqlserverteam.com/axis-order-in-spatial-reference-systems/) which describes the change a bit. I tend to agree it's fine to pass through the changed order as is. Us abstracting this could be more trouble than gain potentially.

File: debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlValueConverters.java [hunk 8]
```diff
@@ -505,6 +505,17 @@ protected boolean matches(String upperCaseTypeName, String upperCaseMatch) {
...
     }
 
+    /**
+     * Determine if the uppercase form of a column's type is geometry collection independent of JDBC driver or server version.
+     *
+     * @param upperCaseTypeName the upper case form of the column's {@link Column#typeName() type name}
+     * @return {@code true} if the type is geometry collection
+     */
+    protected boolean isGeometryCollection(String upperCaseTypeName) {
+        if (upperCaseTypeName == null) return false;
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlValueConverters.java]: Can we agree on not using the brace-less `if` form? I know it's used in this class quite a lot, but I'd highly prefer to get a way from this, as it's so easy to produce nasty bugs due to that. In fact I think this should be added to the CS ruleset. We can handle that all separately, I'll just fix this occurence here while merging the PR.

Comment (jpechane) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlValueConverters.java]: @gunnarmorling I am definitely not a proponent of this style - is is just a copy/paste and specialization of `matches` method. I intentionally did not touch the code.

Comment (gunnarmorling) [reply debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlValueConverters.java]: Cool. Filed https://issues.jboss.org/browse/DBZ-1039 for getting rid of this pattern across the code base.

File: debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/DatabaseDifferences.java [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+ */
+package io.debezium.connector.mysql;
+
+/**
+ * A centralized expression of differences in behaviour between MySQL 5.x and 8.x
+ *
+ * @author Jiri Pechanec
+ *
+ */
+public interface DatabaseDifferences {
```
Comment (gunnarmorling) [inline debezium-connector-mysql/src/test/java/io/debezium/connector/mysql/DatabaseDifferences.java]: Nice idea! I don't think so that methods specific to a single test should be here. I'll hence move the last three which are related only to the single geo-spetial test class and its methods to a separate one.

**Comment by gunnarmorling** [comment]
I've pushed one commit for some clean-up. Will merge it now, let's see whether that test failure in `BinlogReaderBufferIT` shows u on CI, too. Thanks, @jpechane.
